File: seiseifu_research_tool.py
import json
import os
import sys
from datetime import datetime

class Color:
    HEADER = '\033[95m'
    BLUE = '\033[94m'
    CYAN = '\033[96m'
    GREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'

class NanbokuchoAdvancedSearch:
    def __init__(self, data_file='kyushu_nanbokucho_data.json'):
        self.data_file = data_file
        self.data = self.load_data()
        self.history = []

    def load_data(self):
        if not os.path.exists(self.data_file):
            print(f"{Color.FAIL}データファイルが見つかりません: {self.data_file}{Color.ENDC}")
            return {}
        with open(self.data_file, 'r', encoding='utf-8') as f:
            return json.load(f)

    def save_data(self):
        with open(self.data_file, 'w', encoding='utf-8') as f:
            json.dump(self.data, f, ensure_ascii=False, indent=4)

    def search(self, query):
        results = {
            'figures': [],
            'events': [],
            'locations': []
        }
        query = query.lower()
        
        # 人物検索
        for fig in self.data.get('figures', []):
            if query in fig['name'].lower() or query in fig['role'].lower() or query in fig.get('faction', '').lower():
                results['figures'].append(fig)
        
        # 事件検索
        for ev in self.data.get('events', []):
            if query in ev['name'].lower() or query in ev.get('description', '').lower():
                results['events'].append(ev)

        # 場所検索
        for loc in self.data.get('key_locations', []):
            if query in loc['name'].lower() or query in loc.get('description', '').lower():
                results['locations'].append(loc)
        
        return results

    def display_results(self, results):
        if results['figures']:
            print(f"\n{Color.HEADER}=== 人物 ==={Color.ENDC}")
            for i, fig in enumerate(results['figures'], 1):
                print(f"{i}. {Color.CYAN}{fig['name']}{Color.ENDC} ({fig['role']}) - {fig['faction']}")
        
        if results['events']:
            print(f"\n{Color.HEADER}=== 事件 ==={Color.ENDC}")
            for i, ev in enumerate(results['events'], 1):
                year = ev.get('year', '不明')
                print(f"{i}. {Color.GREEN}{year}年: {ev['name']}{Color.ENDC}")

        if results['locations']:
            print(f"\n{Color.HEADER}=== 地点 ==={Color.ENDC}")
            for i, loc in enumerate(results['locations'], 1):
                print(f"{i}. {Color.BLUE}{loc['name']}{Color.ENDC} ({loc.get('terrain', '不明')})")

    def add_memo(self, topic, memo_text):
        if 'memos' not in self.data:
            self.data['memos'] = []
        
        self.data['memos'].append({
            'topic': topic,
            'content': memo_text,
            'timestamp': datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        })
        self.save_data()
        print(f"{Color.GREEN}メモを保存しました。{Color.ENDC}")

    def show_memos(self):
        if 'memos' not in self.data or not self.data['memos']:
            print("メモはありません。")
            return
        print(f"\n{Color.BOLD}=== 研究メモ一覧 ==={Color.ENDC}")
        for i, m in enumerate(self.data['memos'], 1):
            print(f"{i}. [{m['timestamp']}] {Color.CYAN}{m['topic']}{Color.ENDC}: {m['content']}")

def main():
    searcher = NanbokuchoAdvancedSearch()
    print(f"{Color.BOLD}{Color.HEADER}南北朝・征西府 高度研究ツール v1.0{Color.ENDC}")
    
    while True:
        print("\n[1] 検索  [2] メモ追加  [3] メモ表示  [4] データ更新(仮)  [q] 終了")
        choice = input("選択してください > ").strip().lower()
        
        if choice == '1':
            q = input("検索キーワードを入力 > ").strip()
            if not q: continue
            res = searcher.search(q)
            searcher.display_results(res)
        elif choice == '2':
            topic = input("トピック (例: 筑後川の戦い) > ").strip()
            content = input("内容 > ").strip()
            searcher.add_memo(topic, content)
        elif choice == '3':
            searcher.show_memos()
        elif choice == 'q':
            break
        else:
            print("無効な選択です。")

if __name__ == "__main__":
    main()

File: seiseifu_research_tool.zig
const std = @import("std");
const json = std.json;
const fs = std.fs;
const io = std.io;

const Color = struct {
    pub const HEADER = "\x1b[95m";
    pub const BLUE = "\x1b[94m";
    pub const CYAN = "\x1b[96m";
    pub const GREEN = "\x1b[92m";
    pub const WARNING = "\x1b[93m";
    pub const FAIL = "\x1b[91m";
    pub const ENDC = "\x1b[0m";
    pub const BOLD = "\x1b[1m";
};

const Figure = struct {
    name: []const u8,
    faction: []const u8,
    role: []const u8,
};

const Event = struct {
    year: ?u32 = null,
    name: []const u8,
    description: []const u8,
};

const Location = struct {
    name: []const u8,
    terrain: []const u8,
    description: []const u8,
};

const NanbokuchoData = struct {
    title: []const u8,
    figures: []Figure,
    events: []Event,
    key_locations: []Location,
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const stdout = io.getStdOut().writer();
    const stdin = io.getStdIn().reader();

    try stdout.print("{s}{s}南北朝・征西府 高度研究ツール (Zig版) v1.0{s}\n", .{ Color.BOLD, Color.HEADER, Color.ENDC });

    // データの読み込み
    const file = fs.cwd().openFile("kyushu_nanbokucho_data.json", .{}) catch |err| {
        try stdout.print("{s}データファイルが見つかりません: {any}{s}\n", .{ Color.FAIL, err, Color.ENDC });
        return;
    };
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    const parsed = try json.parseFromSlice(NanbokuchoData, allocator, content, .{ .ignore_unknown_fields = true });
    defer parsed.deinit();
    const data = parsed.value;

    var buf: [100]u8 = undefined;

    while (true) {
        try stdout.print("\n[1] 検索  [q] 終了\n選択してください > ", .{});
        const input = (try stdin.readUntilDelimiterOrEof(&buf, '\n')) orelse break;
        const choice = std.mem.trim(u8, input, " \r\n");

        if (std.mem.eql(u8, choice, "q")) break;

        if (std.mem.eql(u8, choice, "1")) {
            try stdout.print("検索キーワードを入力 > ", .{});
            const q_input = (try stdin.readUntilDelimiterOrEof(&buf, '\n')) orelse continue;
            const query = std.mem.trim(u8, q_input, " \r\n");

            if (query.len == 0) continue;

            // 人物検索
            var found_fig = false;
            for (data.figures) |fig| {
                if (std.ascii.indexOfIgnoreCase(fig.name, query) != null or
                    std.ascii.indexOfIgnoreCase(fig.role, query) != null) {
                    if (!found_fig) {
                        try stdout.print("\n{s}=== 人物 ==={s}\n", .{ Color.HEADER, Color.ENDC });
                        found_fig = true;
                    }
                    try stdout.print("- {s}{s}{s} ({s}) - {s}\n", .{ Color.CYAN, fig.name, Color.ENDC, fig.role, fig.faction });
                }
            }

            // 事件検索
            var found_event = false;
            for (data.events) |ev| {
                if (std.ascii.indexOfIgnoreCase(ev.name, query) != null or
                    std.ascii.indexOfIgnoreCase(ev.description, query) != null) {
                    if (!found_event) {
                        try stdout.print("\n{s}=== 事件 ==={s}\n", .{ Color.HEADER, Color.ENDC });
                        found_event = true;
                    }
                    const year = if (ev.year) |y| try std.fmt.allocPrint(allocator, "{}", .{y}) else try allocator.dupe(u8, "不明");
                    defer allocator.free(year);
                    try stdout.print("- {s}{s}年: {s}{s}\n", .{ Color.GREEN, year, ev.name, Color.ENDC });
                }
            }

            // 地点検索
            var found_loc = false;
            for (data.key_locations) |loc| {
                if (std.ascii.indexOfIgnoreCase(loc.name, query) != null or
                    std.ascii.indexOfIgnoreCase(loc.description, query) != null) {
                    if (!found_loc) {
                        try stdout.print("\n{s}=== 地点 ==={s}\n", .{ Color.HEADER, Color.ENDC });
                        found_loc = true;
                    }
                    try stdout.print("- {s}{s}{s} ({s})\n", .{ Color.BLUE, loc.name, Color.ENDC, loc.terrain });
                }
            }
        }
    }
}
